extensions/s3/src/main/java/io/deephaven/extensions/s3/BufferPool.java

```diff
@@ -1,28 +1,34 @@
 package io.deephaven.extensions.s3;
 
-import io.deephaven.util.SafeCloseable;
-import org.jetbrains.annotations.Nullable;
+import io.deephaven.util.datastructures.SegmentedSoftPool;
 
 import java.nio.ByteBuffer;
 
-interface BufferPool {
+final class BufferPool {
 
-    interface BufferHolder extends SafeCloseable {
+    private static final int POOL_SEGMENT_CAPACITY = 10;
+    private final SegmentedSoftPool<ByteBuffer> pool;
+    private final int bufferSize;
 
-        /**
-         * @return The buffer if available, else {@code null}
-         */
-        @Nullable
-        ByteBuffer get();
+    /**
+     * @param bufferSize Upper limit on size of buffers to be pooled
+     */
+    BufferPool(final int bufferSize) {
+        this.bufferSize = bufferSize;
+        this.pool = new SegmentedSoftPool<>(
+                POOL_SEGMENT_CAPACITY,
+                () -> ByteBuffer.allocate(bufferSize),
+                ByteBuffer::clear);
+    }
 
-        /**
-         * Return the held buffer to its pool, and cause subsequent calls to {@link #get()} to return {@code null}
-         */
-        void close();
+    public ByteBuffer take(final int size) {
```

Comment:
I am still somewhat inclined to push for a reference-counted `PooledReference<ByteBuffer>` or similar, and removing the `give` method from this class. This let's us guarantee safety when code uses the proper acquire and release pattern.